# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/clip.rs
# Commit: b04340bad99d
# Full Hash: b04340bad99d70638efda2e7d528e89df77e47c4
# Author: Glenn Watson <git@intuitionlibrary.com>
# Date: 2020-05-22 09:43:16
# Description:
#   Bug 1617524 - Fix crash in get_relative_transform edge case. r=jnicol
#   
#   Previously, WR would attempt to detect at the start of frame
#   building if the spatial node of any picture cache contained
#   a non-axis-aligned transform, and disable picture caching in
# ==============================================================================

diff -r fdfb16cb82a6 -r b04340bad99d gfx/wr/webrender/src/clip.rs
--- a/gfx/wr/webrender/src/clip.rs	Thu May 21 18:23:11 2020 +0000
+++ b/gfx/wr/webrender/src/clip.rs	Thu May 21 07:42:32 2020 +0000
@@ -98,7 +98,7 @@
 use api::image_tiling::{self, Repetition};
 use crate::border::{ensure_no_corner_overlap, BorderRadiusAu};
 use crate::box_shadow::{BLUR_SAMPLE_SCALE, BoxShadowClipSource, BoxShadowCacheKey};
-use crate::spatial_tree::{ROOT_SPATIAL_NODE_INDEX, SpatialTree, SpatialNodeIndex};
+use crate::spatial_tree::{ROOT_SPATIAL_NODE_INDEX, SpatialTree, SpatialNodeIndex, CoordinateSystemId};
 use crate::ellipse::Ellipse;
 use crate::gpu_cache::{GpuCache, GpuCacheHandle, ToGpuBlocks};
 use crate::gpu_types::{BoxShadowStretchMode};
@@ -884,8 +884,27 @@
     /// stack of clips to be propagated.
     pub fn push_surface(
         &mut self,
-        shared_clips: &[ClipInstance],
+        maybe_shared_clips: &[ClipInstance],
+        spatial_tree: &SpatialTree,
     ) {
+        let mut shared_clips = Vec::new();
+
+        // If there are clips in the shared list for a picture cache, only include
+        // them if they are simple, axis-aligned clips (i.e. in the root coordinate
+        // system). This is necessary since when compositing picture cache tiles
+        // into the parent, we don't support applying a clip mask. This only ever
+        // occurs in wrench tests, not in display lists supplied by Gecko.
+        // TODO(gw): We can remove this when we update the WR API to have better
+        //           knowledge of what coordinate system a clip must be in (by
+        //           knowing if a reference frame exists in the chain between the
+        //           clip's spatial node and the picture cache reference spatial node).
+        for clip in maybe_shared_clips {
+            let spatial_node = &spatial_tree.spatial_nodes[clip.spatial_node_index.0 as usize];
+            if spatial_node.coordinate_system_id == CoordinateSystemId::root() {
+                shared_clips.push(*clip);
+            }
+        }
+
         let level = ClipChainLevel {
             shared_clips: shared_clips.to_vec(),
             first_clip_index: self.clips.len(),